### PR TITLE
debuginfo: add compiler option to allow compressed debuginfo sections

### PR DESCRIPTION
**Pull request by durin42**
LLVM already supports emitting compressed debuginfo. In debuginfo=full builds, the debug section is often a large amount of data, and it typically compresses very well (3x is not unreasonable.) We add a new knob to allow debuginfo to be compressed when the matching LLVM functionality is present. Like clang, if a known-but-disabled compression mechanism is requested, we disable compression and emit uncompressed debuginfo sections.

The API is different enough on older LLVMs we just pretend the support
is missing on LLVM older than 16.

### CODE REVIEW

**Comment by durin42** [comment]
This failure is what I'd expect given no way to detect missing zlib
support, so maybe someone has an idea how we should check that in the test?

**Comment by oli-obk** [comment]
is this failure because our CI does not have zlib installed?

**Comment by durin42** [comment]
That's my assumption - I'm not familiar enough with the Rust project's LLVM setup to know how to check.

I'm also not sure how to make the test gracefully degrade if zlib is missing. :(

**Comment by durin42** [comment]
Right - I'm not sure how I should be printing a warning from where I've got access to the LLVM API in question. Any pointers?

Alternatively, if I can run `ld.lld` in the test, I can check what compression methods it supports.

**Comment by durin42** [comment]
Oh, I might have had the dumb. I may see what I missed on my last attempt to print the warning...

**Comment by durin42** [comment]
Ah, I remember the stumbling block I hit: rustc_codegen_llvm isn't currently used by rustc_session, and that seems like a layering violation to me naively, so I can't call into the FFI functions to look if the requested compression is available from there. But by the time we're in a location where we _can_ check that in rustc_codegen_llvm, I don't seem to have a way to emit warnings at all (eg from `target_machine_factory`) - so I'm stumped where I should do this - if there's an example on how to emit the warning based on the codegen backend, I'm sadly missing it. :(

Any hints?

**Comment by durin42** [comment]
Manishearth got me sorted as far as I can tell on printing the diagnostic, and I came up with some Shell Crimes for the test in the makefile, so I think this is ready for review.

File: compiler/rustc_session/src/options.rs [hunk 1]
```diff
@@ -1422,7 +1437,9 @@ options! {
         "inject the given attribute in the crate"),
     debug_info_for_profiling: bool = (false, parse_bool, [TRACKED],
         "emit discriminators and other data necessary for AutoFDO"),
-    debug_macros: bool = (false, parse_bool, [TRACKED],
+        debuginfo_compression: DebugInfoCompression = (DebugInfoCompression::None, parse_debuginfo_compression, [TRACKED],
+            "compress debug info sections (none, zlib, zstd, default: none)"),
+        debug_macros: bool = (false, parse_bool, [TRACKED],
```
Comment (oli-obk) [inline compiler/rustc_session/src/options.rs:1442]: something went wrong with the indentation here

**Review by oli-obk — CHANGES_REQUESTED** [review]
just a style nit, then this lgtm

File: compiler/rustc_session/src/options.rs (hunk 1 above)
Comment (durin42) [reply compiler/rustc_session/src/options.rs]: I feel betrayed by rustfmt.

Fixed!

**Comment by fmease** [comment]
@bors r-
https://github.com/rust-lang/rust/pull/115606#issuecomment-1708484334

**Comment by Wyvern** [comment]
Just tried this option got ```warning: unknown debuginfo compression algorithm zstd - will fall back to uncompressed debuginfo```, how to use `zstd` as compression method?

**Comment by durin42** [comment]
That means the LLVM your rustc was compiled against didn't have zstandard support enabled. zlib is probably already on, maybe try that?

**Pull request merged**